--- ext-2.0.2-src/source/widgets/layout/TableLayout.js	2008-06-17 09:17:59.000000000 +0100
+++ /home/sgodden/products/ext-2.0.2/source/widgets/layout/TableLayout.js	2008-06-19 10:25:22.000000000 +0100
@@ -87,6 +87,14 @@
             this.table = target.createChild(
                 {tag:'table', cls:'x-table-layout', cellspacing: 0, cn: {tag: 'tbody'}}, null, true);
 
+            if (this.tableStyle) {
+                Ext.DomHelper.applyStyles(this.table, this.tableStyle);
+            }
+
+            if (this.tableBorder) {
+                this.table.border = "1";
+            }
+
             this.renderAll(ct, target);
         }
     },
@@ -131,6 +139,12 @@
                 if (c.cellStyle) {
                         Ext.DomHelper.applyStyles(td, c.cellStyle);
                 }
+                if (c.cellAlign) {
+                        td.align = c.cellAlign;
+                }
+                if (c.cellVAlign) {
+                        td.vAlign = c.cellVAlign;
+                }
 		this.getRow(curRow).appendChild(td);
 		return td;
 	},
