--- ext-all-debug.js	2009-02-06 14:49:15.000000000 +0000
+++ ext-all-debug.js	2009-02-06 14:52:39.000000000 +0000
@@ -15152,8 +15152,27 @@
                 this[pos].render(target, c);
             }
             this.rendered = true;
-        }
+        } else {
+            var items = ct.items.items;
+            collapsed = [];
 
+            for(var i = 0, len = items.length; i < len; i++) {
+                var c = items[i];
+                if (!c.rendered) {
+                    var pos = c.region;
+                    if(c.collapsed){
+                        collapsed.push(c);
+                    }
+                    c.collapsed = false;
+                    c.cls = c.cls ? c.cls +' x-border-panel' : 'x-border-panel';
+                    c.render(target, i);
+                    this[pos] = pos != 'center' && c.split ?
+                        new Ext.layout.BorderLayout.SplitRegion(this, c.initialConfig, pos) :
+                    this[pos].render(target, c);
+                }
+            }
+        }
+        
         var size = target.getViewSize();
         if(size.width < 20 || size.height < 20){ // display none?
             if(collapsed){
