--- ext-all-debug.js	2009-03-13 10:12:48.000000000 +0000
+++ ext-all-debug.js	2009-03-13 11:24:34.000000000 +0000
@@ -15084,11 +15084,31 @@
         }
         this.renderAll(ct, this.innerCt);
 
-        var size = Ext.isIE && target.dom != Ext.getBody().dom ? target.getStyleSize() : target.getViewSize();
+        var size = target.dom != Ext.getBody().dom ? target.getStyleSize() : target.getViewSize();
 
         if(size.width < 1 && size.height < 1){ // display none?
             return;
         }
+        
+        // Firefox reports 100% width un-rendered components as having width 5000
+        if (size.width == 5000 && this.container != null) {
+            var thisContainer = this.container;
+            var marginsAndPads = 0;
+            do {
+                if (thisContainer.el.dom != Ext.getBody().dom) {
+                    size = thisContainer.el.getStyleSize();
+                    marginsAndPads += thisContainer.el.getMargins().left;
+                    marginsAndPads += thisContainer.el.getMargins().right;
+                    marginsAndPads += thisContainer.el.getPadding('lr');
+                } else {
+                    Ext.getBody().getViewSize();
+                }
+                thisContainer = thisContainer.ownerCt;
+            } while (size.width == 5000 && thisContainer != null);
+            size.width = size.width - marginsAndPads;
+        } else if (size.width == 5000) {
+            size = Ext.getBody().getViewSize();
+        }
 
         var w = size.width - target.getPadding('lr') - this.scrollOffset,
             h = size.height - target.getPadding('tb'),
