--- ext-all-debug.js	2013-03-25 14:47:26.808197324 +0000
+++ ext-all-debug.js	2013-03-25 14:47:58.984069602 +0000
@@ -28519,7 +28519,7 @@
 
     setValue : function(v){
       v = typeof v == 'number' ? v : parseFloat(String(v).replace(this.decimalSeparator, "."));
-        v = isNaN(v) ? '' : String(v).replace(".", this.decimalSeparator);
+      v = isNaN(v) ? '' : parseFloat(String(v).replace(".", this.decimalSeparator));
         Ext.form.NumberField.superclass.setValue.call(this, v);
     },
 
