=== modified file 'src/client/echo/Application.js'
--- src/client/echo/Application.js	2009-01-02 10:23:54 +0000
+++ src/client/echo/Application.js	2009-02-06 20:43:28 +0000
@@ -1178,7 +1178,7 @@
     set: function(name, newValue) {
         var oldValue = this._localStyle[name];
         if (oldValue === newValue) {
-            return;
+            //return;
         }
         this._localStyle[name] = newValue;
         if (this._listenerList && this._listenerList.hasListeners("property")) {

