
Property changes on: .
___________________________________________________________________
Name: svn:ignore
   + .classpath
.project
.settings


Index: src/server-java/webcontainer/nextapp/echo/webcontainer/resource/RemoteClient.js
===================================================================
--- src/server-java/webcontainer/nextapp/echo/webcontainer/resource/RemoteClient.js	(revision 1087)
+++ src/server-java/webcontainer/nextapp/echo/webcontainer/resource/RemoteClient.js	(working copy)
@@ -31,6 +31,13 @@
     },
     
     /**
+     *
+     * @private
+     * @type Core.ListenerList
+     */
+    _listenerList: null,
+    
+    /**
      * The base server url.
      * @type String
      * @private
@@ -129,6 +136,7 @@
         EchoClient.call(this);
         
         this._serverUrl = serverUrl;
+        this._listenerList = new Core.ListenerList();
         this._processClientUpdateRef = Core.method(this, this._processClientUpdate);
         this._processClientEventRef = Core.method(this, this._processClientEvent);
         this._urlMappings = {};
@@ -155,6 +163,17 @@
     },
     
     /**
+     * Adds a listener to be notified when a server update has been
+     * completely processed, that is, all component and rendering
+     * updates have been completed.
+     * 
+     * @param {Function} l the listener to add
+     */
+    addServerUpdateCompleteListener: function(l) {
+        this._listenerList.addListener("serverUpdateComplete", l);
+    },
+    
+    /**
      * Decompresses a shorthand URL into a valid full-length URL.
      * A shorthand URL is expressed as "!A!xxxx" where
      * "A" is a key whose value contains the first portion of the URL
@@ -354,6 +373,10 @@
             this._waitIndicatorActive = false;
             this._waitIndicator.deactivate();
         }
+        
+        if (this._listenerList.hasListeners("serverUpdateComplete")) {
+            this._listenerList.fireEvent({type: "serverUpdateComplete"});
+        }
     },
     
     /**
@@ -417,8 +440,17 @@
     removeComponentListener: function(component, eventType) {
         component.removeListener(eventType, this._processClientEventRef);
     },
-    
+        
     /**
+     * Removes a ServerUpdateCompleteListener.
+     * 
+     * @param {Function} l the listener to remove
+     */
+    removeServerUpdateCompleteListener: function(l) {
+        this._listenerList.removeListener("serverUpdateComplete", l);
+    },
+
+    /**
      * Sets the wait indicator that will be displayed when a client-server action takes longer than
      * a specified period of time.
      * 
Index: src/server-java/webcontainer/nextapp/echo/webcontainer/service/WindowHtmlService.java
===================================================================
--- src/server-java/webcontainer/nextapp/echo/webcontainer/service/WindowHtmlService.java	(revision 1087)
+++ src/server-java/webcontainer/nextapp/echo/webcontainer/service/WindowHtmlService.java	(working copy)
@@ -105,6 +105,13 @@
         styleElement.setAttribute("type", "text/css");
         styleElement.appendChild(document.createTextNode(" "));
         headElement.appendChild(styleElement);
+        
+        Element extStyleElement = document.createElement("link");
+        extStyleElement.setAttribute("href", "resources/css/ext-all.css");
+        extStyleElement.setAttribute("type", "text/css");
+        extStyleElement.setAttribute("rel", "stylesheet");
+        extStyleElement.appendChild(document.createTextNode(" "));
+        headElement.appendChild(extStyleElement);
 
         Element scriptElement = document.createElement("script");
         Text textNode = document.createTextNode(" ");
Index: ant.properties
===================================================================
--- ant.properties	(revision 1087)
+++ ant.properties	(working copy)
@@ -3,7 +3,7 @@
 ant.build.javac.source 1.4
 ant.build.javac.target 1.4
 
-servlet.lib.jar                         ${env.SERVLET_LIB_JAR}
+servlet.lib.jar                         /Program Files/Apache Software Foundation/Apache Tomcat 6.0.14/lib/servlet-api.jar
 servlet.lib.msg                         The environment variable SERVLET_LIB_JAR must contain the path to the Servlet 2.4 \
                                         specification JAR file (servlet.jar or servlet-api.jar).
 debug                                   yes
