--- ext-all-debug.js	2009-02-06 14:39:48.000000000 +0000
+++ ext-all-debug.js	2009-02-06 14:41:00.000000000 +0000
@@ -12563,7 +12563,9 @@
             if(this.baseParams){
                 delete this.baseParams.groupBy;
             }
-            this.reload();
+            var params = { groupBy: null };
+            var options = {params: params};
+            this.reload(options);
         }else{
             this.applySort();
             this.fireEvent('datachanged', this);
