diff -r a98f688169ff src/client/echo/Application.js
--- a/src/client/echo/Application.js	Mon Jul 05 13:18:33 2010 +0100
+++ b/src/client/echo/Application.js	Thu Sep 30 15:40:00 2010 +0100
@@ -603,7 +603,25 @@
          *  Flag indicating whether component is rendered as a pane (pane components consume available height).
          *  @type Boolean 
          */
-        pane: false
+        pane: false,
+        
+        /**
+         * Method that may focus the next component when requested. Returns
+         * the component if the next focus is set to a new component, returns true
+         * to skip the focus setting.
+         */
+        focusNextComponent: function() {
+        	return null;
+        },
+        
+        /**
+         * Method that may focus the previous component when requested. Returns
+         * the component if the previous focus is set to a new component, returns true
+         * to skip the focus setting.
+         */
+        focusPrevComponent: function() {
+        	return null;
+        }
     },
     
     /**
@@ -1416,6 +1434,23 @@
             }
         }
         
+        // ask the component to focus the next/previous component
+        if (reverse) {
+        	var pc = component.focusPrevComponent();
+        	if (pc != null) {
+        		return pc;
+        	} else if (pc === true) {
+        		return null;
+        	}
+        } else {
+        	var pc = component.focusNextComponent();
+        	if (pc != null) {
+        		return pc;
+        	} else if (pc === true) {
+        		return null;
+        	}
+        }
+        
         // If a specific next focusable component has been specified, attempt to focus it.
         var setComponentId = reverse ? component.focusPreviousId : component.focusNextId;
         if (setComponentId) {
