diff -r 9d7a787154b6 src/server-java/app/nextapp/echo/app/Window.java
--- a/src/server-java/app/nextapp/echo/app/Window.java	Wed Oct 28 10:18:46 2009 +0000
+++ b/src/server-java/app/nextapp/echo/app/Window.java	Thu Nov 05 09:44:10 2009 +0000
@@ -118,10 +118,10 @@
     private Map renderIdToComponentMap;
     
     /**
-     * A list of components removed in the current request
+     * A map of component ids to component instances removed in the current request
      * @return
      */
-    private List componentsToRemove;
+    private Map componentsToRemove;
     
     /**
      * The current transactionId.  Used to ensure incoming ClientMessages reflect
@@ -657,9 +657,9 @@
      */
     void unregisterComponent(Component component) {
     	if (componentsToRemove == null) {
-    		componentsToRemove = new LinkedList();
+    		componentsToRemove = new HashMap();
     	}
-    	componentsToRemove.add(component);
+    	componentsToRemove.put(component.getRenderId(), component);
     }
     
     /**
@@ -667,13 +667,15 @@
      */
     public void processComponentRemovals() {
     	if (componentsToRemove != null) {
-    		Iterator i = componentsToRemove.iterator();
+    		Iterator i = componentsToRemove.entrySet().iterator();
     		while (i.hasNext()) {
-    			Component component = (Component)i.next();
-    	        renderIdToComponentMap.remove(component.getRenderId());
-    	        if (component instanceof ModalSupport && ((ModalSupport) component).isModal()) {
-    	            setModal(component, false);
-    	        }
+    		    Map.Entry entry = (Map.Entry)i.next();
+    		    String renderId = (String)entry.getKey();
+    			//Component component = (Component)entry.getValue();
+    	        renderIdToComponentMap.remove(renderId);
+    	        //if (component instanceof ModalSupport && ((ModalSupport) component).isModal()) {
+    	        //    setModal(component, false);
+    	        //}
     		}
     		componentsToRemove = null;
     	}
