diff -r b10ddd734761 src/client/echo/Client.js
--- a/src/client/echo/Client.js	Tue Dec 01 12:57:49 2009 +0000
+++ b/src/client/echo/Client.js	Tue Jan 05 11:46:21 2010 +0000
@@ -482,7 +482,12 @@
             this.dispose();
         } finally {
             if (this.configuration["StopError.URI"]) {
-                var eid = detail.substring(detail.indexOf("ID: ") + 4);
+                var eid = '';
+                if (detail != null && detail.indexOf("Server Exception. ID: ") > -1) {
+                    eid = detail.substring(detail.indexOf("ID: ") + 4);
+                } else if (detail != null) {
+                    eid = detail;
+                }
                 var url = this.configuration["StopError.URI"];
                 if (url.indexOf('?') > -1) {
                     url = url + "&eid=" + eid;
