Index: src/server-java/app/META-INF/nextapp/echo/XmlPeers.properties
===================================================================
--- src/server-java/app/META-INF/nextapp/echo/XmlPeers.properties	(revision 1277)
+++ src/server-java/app/META-INF/nextapp/echo/XmlPeers.properties	(working copy)
@@ -16,6 +16,7 @@
 nextapp.echo.app.Insets                      nextapp.echo.app.serial.property.InsetsPeer
 nextapp.echo.app.LayoutData                  nextapp.echo.app.serial.property.LayoutDataPeer
 nextapp.echo.app.ResourceImageReference      nextapp.echo.app.serial.property.ResourceImageReferencePeer
+nextapp.echo.app.HttpImageReference          nextapp.echo.webcontainer.sync.property.HttpImageReferencePeer
 
 nextapp.echo.app.layout.CellLayoutData       nextapp.echo.app.serial.property.CellLayoutDataPeer
 nextapp.echo.app.layout.ColumnLayoutData     nextapp.echo.app.serial.property.ColumnLayoutDataPeer
Index: src/server-java/webcontainer/nextapp/echo/webcontainer/resource/RemoteClient.js
===================================================================
--- src/server-java/webcontainer/nextapp/echo/webcontainer/resource/RemoteClient.js	(revision 1277)
+++ src/server-java/webcontainer/nextapp/echo/webcontainer/resource/RemoteClient.js	(working copy)
@@ -46,6 +46,13 @@
     },
     
     /**
+     * Listeners to be notified of remote client events.
+     * @private
+     * @type Core.ListenerList
+     */
+    _listenerList: null,
+    
+    /**
      * The base server url.
      * @type String
      * @private
@@ -144,6 +151,7 @@
         Echo.Client.call(this);
         
         this._serverUrl = serverUrl;
+        this._listenerList = new Core.ListenerList();
         this._processClientUpdateRef = Core.method(this, this._processClientUpdate);
         this._processClientEventRef = Core.method(this, this._processClientEvent);
         this._urlMappings = {};
@@ -170,6 +178,17 @@
     },
     
     /**
+     * Adds a listener to be notified when a server update has been
+     * completely processed, that is, all component and rendering
+     * updates have been completed.
+     * 
+     * @param {Function} l the listener to add
+     */
+    addServerUpdateCompleteListener: function(l) {
+        this._listenerList.addListener("serverUpdateComplete", l);
+    },
+    
+    /**
      * Decompresses a shorthand URL into a valid full-length URL.
      * A shorthand URL is expressed as "!A!xxxx" where
      * "A" is a key whose value contains the first portion of the URL
@@ -384,6 +403,10 @@
             this._waitIndicatorActive = false;
             this._waitIndicator.deactivate();
         }
+        
+        if (this._listenerList.hasListeners("serverUpdateComplete")) {
+            this._listenerList.fireEvent({type: "serverUpdateComplete"});
+        }
     },
     
     /**
@@ -439,8 +462,17 @@
     removeComponentListener: function(component, eventType) {
         component.removeListener(eventType, this._processClientEventRef);
     },
-    
+        
     /**
+     * Removes a ServerUpdateCompleteListener.
+     * 
+     * @param {Function} l the listener to remove
+     */
+    removeServerUpdateCompleteListener: function(l) {
+        this._listenerList.removeListener("serverUpdateComplete", l);
+    },
+
+    /**
      * Sets the wait indicator that will be displayed when a client-server action takes longer than
      * a specified period of time.
      * 
Index: src/server-java/webcontainer/nextapp/echo/webcontainer/WebContainerServlet.java
===================================================================
--- src/server-java/webcontainer/nextapp/echo/webcontainer/WebContainerServlet.java	(revision 1277)
+++ src/server-java/webcontainer/nextapp/echo/webcontainer/WebContainerServlet.java	(working copy)
@@ -41,9 +41,11 @@
 
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 
+import java.util.Set;
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
@@ -388,5 +390,22 @@
             activeConnection.set(null);
         }
     }
+   
+    private Set cssFileNames = new HashSet();
     
-}
+    /**
+     * Adds a css file name to be added to the window html.
+     */
+    public void addCssFileName(String cssFileName) {
+        cssFileNames.add(cssFileName);
+    }
+    
+    /**
+     * Returns an iterator over the CSS file names.
+     * @return the iterator.
+     */
+    public Iterator getCssFileNames() {
+        return cssFileNames.iterator();
+    }
+    
+}
\ No newline at end of file
Index: src/server-java/webcontainer/nextapp/echo/webcontainer/service/WindowHtmlService.java
===================================================================
--- src/server-java/webcontainer/nextapp/echo/webcontainer/service/WindowHtmlService.java	(revision 1277)
+++ src/server-java/webcontainer/nextapp/echo/webcontainer/service/WindowHtmlService.java	(working copy)
@@ -104,6 +104,21 @@
         styleElement.setAttribute("type", "text/css");
         styleElement.appendChild(document.createTextNode(" "));
         headElement.appendChild(styleElement);
+        
+        WebContainerServlet servlet = conn.getServlet();
+        
+        Iterator cssIt = servlet.getCssFileNames();
+        if (cssIt != null) {
+            while(cssIt.hasNext()) {
+                String cssFileName = (String) cssIt.next();
+                Element additionalStyleElement = document.createElement("link");
+                additionalStyleElement.setAttribute("href", cssFileName);
+                additionalStyleElement.setAttribute("type", "text/css");
+                additionalStyleElement.setAttribute("rel", "stylesheet");
+                additionalStyleElement.appendChild(document.createTextNode(" "));
+                headElement.appendChild(additionalStyleElement);
+            }
+        }
 
         Element scriptElement = document.createElement("script");
         Text textNode = document.createTextNode(" ");
@@ -111,8 +126,8 @@
         scriptElement.setAttribute("type", "text/javascript");
         scriptElement.setAttribute("src", userInstance.getServiceUri(BootService.SERVICE));
         headElement.appendChild(scriptElement);
+
         
-        WebContainerServlet servlet = conn.getServlet();
         Iterator scriptIt = servlet.getStartupScripts();
         if (scriptIt != null) {
             while (scriptIt.hasNext()) {
@@ -130,7 +145,7 @@
         bodyElement.setAttribute("id", "body");
         bodyElement.setAttribute("onload", "Echo.Boot.boot('" + userInstance.getServletUri() + "', " + debug + ");");
         bodyElement.setAttribute("style",
-                "height:100%;width:100%;margin:0px;padding: 0px;" +
+                "height:100%;width:100%;overflow:hidden;margin:0px;padding: 0px;" +
                 "font-family:verdana, arial, helvetica, sans-serif;font-size:10pt");
         htmlElement.appendChild(bodyElement);
 
Index: src/server-java/webcontainer/nextapp/echo/webcontainer/util/IdTable.java
===================================================================
--- src/server-java/webcontainer/nextapp/echo/webcontainer/util/IdTable.java	(revision 1277)
+++ src/server-java/webcontainer/nextapp/echo/webcontainer/util/IdTable.java	(working copy)
@@ -101,6 +101,15 @@
      * <code>IdTable</code>.
      */
     private void purge() {
+        /*
+         * Due to clustering, our transient fields may be null.
+         */
+        if (referenceQueue == null) {
+            referenceQueue = new ReferenceQueue();
+        }
+        if (idToReferenceMap == null) {
+            idToReferenceMap = new HashMap();
+        }
         // Convert any hard references to weak references.
         if (hasHardReferences) {
             synchronized (idToReferenceMap) {
Index: src/client/corejs/Core.js
===================================================================
--- src/client/corejs/Core.js	(revision 1277)
+++ src/client/corejs/Core.js	(working copy)
@@ -99,7 +99,7 @@
         var definition = arguments.length == 1 ? arguments[0] : arguments[1];
         
         // Perform argument error checking.
-        if (arguments.length == 2) {
+        if (baseClass) {
             if (typeof(baseClass) != "function") {
                 throw new Error("Base class is not a function, cannot derive.");
             }

Property changes on: src/client/echo
___________________________________________________________________
Name: svn:ignore
   + .Sync.TextComponent.js.swp


Index: src/client/echo/Render.js
===================================================================
--- src/client/echo/Render.js	(revision 1277)
+++ src/client/echo/Render.js	(working copy)
@@ -30,6 +30,23 @@
      * @type Object
      */
     _disposedComponents: null,
+
+        
+    /**
+     * Listeners to be notified of rendering phase events.
+     * @type Core.ListenerList
+     */
+    _listenerList: new Core.ListenerList(),
+
+    /**
+     * Adds a listener to be notified when the renderDisplay
+     * phase of an update has been completed.
+     * 
+     * @param {Function} l the listener to add
+     */
+    addRenderDisplayCompleteListener: function(l) {
+        Echo.Render._listenerList.addListener("renderDisplayComplete", l);
+    },
     
     //FIXME.  Scrollbar position tracking code in SplitPane appears to suggest that
     // disposed states are not in good shape....SplitPane is being disposed when
@@ -262,6 +279,11 @@
                 Echo.Render._doRenderDisplay(updates[i].parent, true);
             }
         }
+
+	// call any listeners that want to know when when the render display phase is finished
+        if (Echo.Render._listenerList.hasListeners("renderDisplayComplete")) {
+            Echo.Render._listenerList.fireEvent({type: "renderDisplayComplete"});
+        }
     
         // Profiling: Mark completion of display phase.
         if (Echo.Client.profilingTimer) {
@@ -298,6 +320,15 @@
         }
         this._peers[componentName] = peerObject;
     },
+
+    /**
+     * Removes a renderDisplayCompleteListener.
+     * 
+     * @param {Function} l the listener to remove
+     */
+    removeRenderDisplayCompleteListener: function(l) {
+        this._listenerList.removeListener("renderDisplayComplete", l);
+    },
     
     /**
      * Renders a new component inside of a DOM element.
Index: src/client/echo/Application.js
===================================================================
--- src/client/echo/Application.js	(revision 1277)
+++ src/client/echo/Application.js	(working copy)
@@ -1047,8 +1047,9 @@
             if (this._style != null) {
                 value = this._style[name];
             }
-            if (value == null && this._styleName && this.application && this.application._styleSheet) {
-                var style = this.application._styleSheet.getRenderStyle(this._styleName, this.componentType);
+            if (value == null && this.application && this.application._styleSheet) {
+                var style = this.application._styleSheet.getRenderStyle(
+                    this._styleName != null ? this._styleName : "", this.componentType);
                 if (style) {
                     value = style[name];
                 }
@@ -1077,7 +1078,8 @@
                 value = valueArray ? valueArray[index] : null;
             }
             if (value == null && this._styleName && this.application && this.application._styleSheet) {
-                var style = this.application._styleSheet.getRenderStyle(this._styleName, this.componentType);
+                var style = this.application._styleSheet.getRenderStyle(
+                    this._styleName != null ? this._styleName : "", this.componentType);
                 if (style) {
                     valueArray = style[name];
                     value = valueArray ? valueArray[index] : null;
